inferno-core/src/Inferno/Core.hs

```diff
@@ -9,30 +9,31 @@ import Control.Monad (foldM)
 import Control.Monad.Catch (MonadCatch, MonadThrow)
 import Control.Monad.Except (MonadFix)
 import Data.Bifunctor (bimap)
-import qualified Data.List.NonEmpty as NEList
+import Data.List.NonEmpty (NonEmpty)
 import qualified Data.Map as Map
 import qualified Data.Set as Set
 import Data.Text (Text)
 import Inferno.Eval (TermEnv, eval, runEvalM)
 import Inferno.Eval.Error (EvalError)
-import Inferno.Infer (inferExpr, inferTypeReps)
+import Inferno.Infer (TypeError, inferExpr, inferTypeReps)
+import Inferno.Infer.Error (Location)
 import Inferno.Infer.Pinned (pinExpr)
 import Inferno.Module (Module (..))
 import Inferno.Module.Builtin (builtinModule)
-import Inferno.Module.Prelude (ModuleMap, baseOpsTable, builtinModulesOpsTable, builtinModulesPinMap, builtinModulesTerms)
-import Inferno.Parse (parseExpr, prettyError)
-import Inferno.Types.Syntax (Expr (App, TypeRep), ExtIdent, SourcePos, collectArrs)
+import Inferno.Module.Prelude (ModuleMap, baseOpsTable, builtinModulesOpsTable, builtinModulesPinMap, builtinModulesTerms, preludeNameToTypeMap)
+import Inferno.Parse (InfernoParsingError, parseExpr)
+import Inferno.Types.Syntax (Comment, CustomType, Expr (App, TypeRep), ExtIdent, ModuleName, Namespace, SourcePos, TypeClass, TypeMetadata, collectArrs)
 import Inferno.Types.Type (ImplType (ImplType), TCScheme (ForallTC))
 import Inferno.Types.Value (ImplEnvM, Value, runImplEnvM)
 import Inferno.Types.VersionControl (Pinned, VCObjectHash, pinnedToMaybe)
 import Inferno.VersionControl.Types (VCObject (VCFunction))
 import Prettyprinter (Pretty)
-import Text.Megaparsec (initialPos)
+import Text.Megaparsec (ParseError, initialPos)
 
 data InfernoError
-  = ParseError String
-  | PinError String
-  | InferenceError String
+  = ParseError (NonEmpty (ParseError Text InfernoParsingError, SourcePos))
+  | PinError [TypeError SourcePos]
+  | InferenceError [TypeError SourcePos]
```

Comment:
Retain type information of different types of errors so that LSP can convert them into the appropriate Diagnostics